platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,208 @@
+
+**We are writing this with the assumption that Github Actions work will continue to completion.**
+
+This phase of the initiative does not include
+- Programmatic isolation of the applications
+- Physical isolation of the applications
+
+### High Level Design
+
+Deployment environments will be remapped from `master` as a central release location to related environment branches. `master` will deploy to production (va.gov), `staging` will deploy to staging (staging.va.gov), and `dev` will deploy to dev (dev.va.gov).
+
+For each environment (prod/staging/dev), a Github Action Workflow will be set up to handle application deployments to the appropriate S3 bucket.
+
+The daily deployment will continue until all applications are approved and migrated into the new production release pattern. The same also applies to any platform or shared code that may need regular deployments.
+
+After the system is rolled out, we will start working from a "disallowed" list instead of an "allowed" list to identify applications that are not yet ready for autonomous deployments (I.E. the static pages app).
+
+VSP will always maintain an option to complete and deploy a full release to any environment as a system of developer hygiene and to keep all dependencies up to date. This can be the current daily deploy script, whether (long term) it is on a schedule or not.
+
+#### Production Deployments 
+
+Commits to the `master` branch will have an optional auto deployment flag  in the manifest.json file (at the discretion of the application, `deploy: true`) as well as a manual option for app teams to control production release with a higher level of granularity. A team may not want to deploy to production until after hours, for example.
+
+Upon initial/beta release of production deployments there will be an "allowed" list for applications that VSP identifies as good candidates for testing the system (teams with frequent stable releases). All other applications will have the advantage of auto deployments to staging and dev environments, but will remain contained in the daily deployment until the system is ready to be truly switched on. Applications using the auto-deployment will also remain in the daily deployment while the system is in beta to clean up and issues and dependencies upgrades. This will also reduce complexity.
+
+#### Staging and Dev deployments
+
+Deployments to staging and dev will be automatic upon merge to the branch for all applications. This is a workflow change and will need heavy communication throughout VFS teams. This will mimic the current staging/dev deployment we have in master currently, but no tests will be run. This will cut down on failures due to flakiness in an environment where it isn't really needed since commits still need to complete the build and run through tests in a PR before merging into master in addition to tests run in master CI.
```

Comment:
At what point the tests will be run? How are we going to make sure the tests passed on staging/dev before going on prod?

Reply:
The merge to staging / dev would have to happen locally, so we can't force any tests to run. I chatted with testing tools team about the e2e and we decided that none of those test would run for staging / dev as well since they only run for prod right now.

We would still have tests running in PR and master, but for the actual isolated deploy, as long as tests have cleared master, they can deploy to prod